[FLINK-21333][coordination] Add StopWithSavepoint state to declarative scheduler

rmetzger opened the pull request:
## What is the purpose of the change

Introduce StopWithSavepoint state to declarative scheduler.

This PR is based on https://github.com/apache/flink/pull/14921.

![image](https://user-images.githubusercontent.com/89049/108170205-1f0cce80-70fa-11eb-83c6-f7a12de21c6f.png)


## Verifying this change

IT and Unit tests.

rmetzger commented:
@tillrohrmann  if you happen to have some time, it would be nice if you could take a quick look at the `StopWithSavepoint` state from the last commit: https://github.com/apache/flink/pull/14948/commits/93af9b697282e16a264142d38a7557a54cb6fb31#diff-9ad4a55dacb2aeda64e2158e01eba31a73b7a89a1cf6b428499f87d7dd63423f
and maybe as well at the first unit test for it: https://github.com/apache/flink/pull/14948/commits/93af9b697282e16a264142d38a7557a54cb6fb31#diff-831a3025bb46e84e3d0244712a0548d8f1fb426ee944bffdf6ba36d6bd1c5201

If time permits, I'll continue working on the tests tonight / tomorrow morning

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/StopWithSavepoint.java:
```diff
@@ -0,0 +1,214 @@
+    private String handleSavepointOperationResult(
+            CompletedCheckpoint completedCheckpoint, Throwable throwable) {
+        if (throwable != null) {
+            context.runIfState(this, () -> handleAnyFailure(throwable));
+            throw new SavepointCreationException(
+                    "Unable to stop with savepoint. Error while creating the savepoint.",
+                    throwable);
```
Handling the failure and sending the result to the user happens here in two places. The latter is even obfuscated a bit through throwing an exception which is forwarded to the result future at some other place. This is a rather complicated control flow. Couldn't we do this in the failure handling logic?

Are we treating savepoint failures as global failures? It looks a bit like this. I think this is wrong. A savepoint failure should only fail the operation and transition us back to the `Executing` state.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/StopWithSavepoint.java:
```diff
@@ -0,0 +1,214 @@
+                                                                        status, path)))
+                        // check that the job finished successfully
+                        .handleAsync(
+                                this::handleStopWithSavepointOperationResult,
```
Why are we handling the final job result by accessing the `getTerminationFuture`? Don't we already have `onGloballyTerminalState` which tells us if the `ExecutionGraph` has reached a globally terminal state?

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/StopWithSavepoint.java:
```diff
@@ -0,0 +1,214 @@
+                final FlinkException exception =
+                        new FlinkException(
+                                "Not all vertices of job are finished. Stop with savepoint operation failed.");
+                context.runIfState(this, () -> handleAnyFailure(exception));
+                throw new CompletionException("Unable to stop with savepoint", exception);
```
here again we have one path for the failure handling and another path for communicating the result back to the user. I think this should be one and the same path. If we leave this state in any other state than `Finished`, then we should fail the operation.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/StopWithSavepoint.java:
```diff
@@ -0,0 +1,214 @@
+    @Override
+    protected void handleAnyFailure(@Nonnull Throwable cause) {
+        Preconditions.checkState(context.isState(this), "Assuming StopWithSavepoint state");
+        // restart the checkpoint coordinator if stopWithSavepoint failed.
+        startCheckpointScheduler(checkpointCoordinator);
```
We could say that we do this whenever we leave this state in any other state than `Finished`.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/StopWithSavepoint.java:
```diff
@@ -0,0 +1,214 @@
+
+    @Override
+    void onGloballyTerminalState(JobStatus globallyTerminalState) {
+        // handled already
```
This should be an indicator that we now have different control paths for one and the same thing. This is usually not ideal and makes maintenance harder.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/StopWithSavepoint.java:
```diff
@@ -0,0 +1,214 @@
+                            "Restarting or Failing job because of failure while stopping with savepoint. Reason: {}",
+                            cause.getMessage());
+        }
+        super.handleAnyFailure(cause);
```
This is a bit of personal taste but I find it hard to see what happens in case of a failure in the state `StopWithSavepoint`. If we didn't use the `ExecutingStateWithFailureHandler` and instead have an explicit `context.goToRestarting` then it would be a bit clearer. It currently works because all failures are handled as global failovers. But once this changes, this class will break because we want to fail with global failovers explicitly.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/StopWithSavepoint.java:
```diff
@@ -0,0 +1,214 @@
+        // will be restarted by the CheckpointCoordinatorDeActivator.
+        checkpointCoordinator.stopCheckpointScheduler();
+
+        CompletableFuture<String> savepointFuture =
+                checkpointCoordinator
+                        .triggerSynchronousSavepoint(advanceToEndOfEventTime, targetDirectory)
+                        // handle errors while creating the savepoint
+                        .handleAsync(
+                                this::handleSavepointOperationResult,
+                                context.getMainThreadExecutor())
+                        // Wait for job to be finished
+                        .thenCompose(
+                                path ->
+                                        executionGraph
+                                                .getTerminationFuture()
+                                                .thenApply(
+                                                        status ->
+                                                                new JobStatusAndSavepointPath(
+                                                                        status, path)))
+                        // check that the job finished successfully
+                        .handleAsync(
+                                this::handleStopWithSavepointOperationResult,
+                                context.getMainThreadExecutor());
+
+        FutureUtils.forward(savepointFuture, operationCompletionFuture);
```
Instead of using future callbacks, the `StopWithSavepoint` state could have small internal state machine which reacts to signals. Two signals we need to handle are the terminal job status and the savepoint status. The former is already available via `onGloballyTerminalState`. The latter could be implemented by having `onSavepointSuccess(String path)` and a `onSavepointFailure(Throwable cause)`. That way we wouldn't have to deal with nested future callbacks. Maybe we could even reuse the improvements from FLINK-21030.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/declarative/StopWithSavepoint.java:
```diff
@@ -0,0 +1,214 @@
+            OperatorCoordinatorHandler operatorCoordinatorHandler,
+            Logger logger,
+            ClassLoader userCodeClassLoader,
+            CheckpointCoordinator checkpointCoordinator,
```
How well is the state testable with accessing the `CheckpointCoordinator` explicitly? Would it be easier if we hid the coordinator behind some interface or a `context` method?

tillrohrmann replied:
E.g. we could have methods for stopping & starting the `CC` and to create a savepoint.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/StopWithSavepointTest.java:
```diff
@@ -0,0 +1,248 @@
+                .getSavepointFuture()
+                .completeExceptionally(new RuntimeException("Savepoint creation failed"));
+
+        stopWithSavepointEnv.getExecutionGraphTerminationFuture().complete(JobStatus.FAILED);
```
Why is this necessary? A savepoint failure should not require the EG to reach a globally terminal state before we can recover?

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/StopWithSavepointTest.java:
```diff
@@ -0,0 +1,248 @@
+        ctx.close(); // trigger outstanding executions
+
+        assertThat(
+                stopWithSavepointEnv.getOperationCompletionFuture().isCompletedExceptionally(),
```
I am not a huge fan of this environment. Looking alone on this test, it is not clear to me why the env has the operation completion future. If the operation future belonged to the state, then it would be clearer, for example.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/StopWithSavepointTest.java:
```diff
@@ -0,0 +1,248 @@
+                savepointFuture);
+    }
+
+    private static class MockCheckpointCoordinator extends CheckpointCoordinator {
```
Let's not continue with this pattern. Extending concrete classes is a recipe for a lot of rabbit holes.

Maybe an easier solution could be to define the interface the `StopWithSavepoint` state requires and then have some adapters to map it to the underlying `CheckpointCoordinator` (or even let the `CheckpointCoordinator` implement this interface).

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/StopWithSavepointTest.java:
```diff
@@ -0,0 +1,248 @@
+
+        MockCheckpointCoordinator(CompletableFuture<CompletedCheckpoint> savepointFuture) {
+            super(
+                    new JobID(),
+                    new CheckpointCoordinatorConfiguration
+                                    .CheckpointCoordinatorConfigurationBuilder()
+                            .build(),
+                    Collections.emptyList(),
+                    new StandaloneCheckpointIDCounter(),
+                    new StandaloneCompletedCheckpointStore(1),
+                    new MemoryStateBackend(),
+                    Executors.directExecutor(),
+                    new CheckpointsCleaner(),
+                    new ManuallyTriggeredScheduledExecutor(),
+                    SharedStateRegistry.DEFAULT_FACTORY,
+                    new CheckpointFailureManager(0, NoOpFailJobCall.INSTANCE),
+                    new CheckpointPlanCalculator(
+                            new JobID(),
+                            Arrays.asList(defaultVertices),
+                            Arrays.asList(defaultVertices),
+                            Arrays.asList(defaultVertices)),
+                    new ExecutionAttemptMappingProvider(Arrays.asList(defaultVertices)));
```
The pain to set these things up should be a good indicator that we are doing something wrong here.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/declarative/StopWithSavepointTest.java:
```diff
@@ -0,0 +1,248 @@
+        }
+    }
+
+    private static class MockExecutionGraph extends ExecutionGraph {
+        private final CompletableFuture<JobStatus> mockTerminationFuture;
+
+        private MockExecutionGraph(CompletableFuture<JobStatus> mockTerminationFuture)
+                throws IOException {
+            super(
+                    new JobInformation(
+                            new JobID(),
+                            "Test Job",
+                            new SerializedValue<>(new ExecutionConfig()),
+                            new Configuration(),
+                            Collections.emptyList(),
+                            Collections.emptyList()),
+                    TestingUtils.defaultExecutor(),
+                    TestingUtils.defaultExecutor(),
+                    AkkaUtils.getDefaultTimeout(),
+                    1,
+                    ExecutionGraph.class.getClassLoader(),
+                    VoidBlobWriter.getInstance(),
+                    PartitionReleaseStrategyFactoryLoader.loadPartitionReleaseStrategyFactory(
+                            new Configuration()),
+                    NettyShuffleMaster.INSTANCE,
+                    NoOpJobMasterPartitionTracker.INSTANCE,
+                    ScheduleMode.EAGER,
+                    NoOpExecutionDeploymentListener.get(),
+                    (execution, newState) -> {},
+                    0L);
+            this.mockTerminationFuture = mockTerminationFuture;
+        }
+
+        @Override
+        public CompletableFuture<JobStatus> getTerminationFuture() {
+            return mockTerminationFuture;
+        }
+    }
```
Yet another `EG` sub class...

tillrohrmann commented on flink-runtime/src/test/resources/log4j2-test.properties:
```diff
@@ -18,7 +18,7 @@
 # Set root logger level to OFF to not flood build logs
 # set manually to INFO for debugging purposes
-rootLogger.level = OFF
+rootLogger.level = INFO
```
Seems like not intended.

tillrohrmann commented on flink-tests/src/test/resources/log4j2-test.properties:
```diff
@@ -18,7 +18,7 @@
 # Set root logger level to OFF to not flood build logs
 # set manually to INFO for debugging purposes
-rootLogger.level = OFF
+rootLogger.level = INFO
```
Probably not intended.

tillrohrmann commented on flink-tests/src/test/java/org/apache/flink/test/scheduling/DeclarativeSchedulerITCase.java:
```diff
@@ -88,6 +103,155 @@ public void testGlobalFailoverCanRecoverState() throws Exception {
+            testStopWithSavepoint(StopWithSavepointTestBehavior.FAIL_ON_CHECKPOINT);
+            fail("Expect exception");
+        } catch (ExecutionException e) {
+            assertThat(e.getMessage(), containsString("CheckpointException"));
```
`FlinkMatchers.containsMessage`

tillrohrmann commented on flink-tests/src/test/java/org/apache/flink/test/scheduling/DeclarativeSchedulerITCase.java:
```diff
@@ -88,6 +103,155 @@ public void testGlobalFailoverCanRecoverState() throws Exception {
+        assumeTrue(ClusterOptions.isDeclarativeResourceManagementEnabled(configuration));
+
+        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+        if (behavior == StopWithSavepointTestBehavior.FAIL_ON_FIRST_CHECKPOINT_ONLY) {
+            env.setRestartStrategy(RestartStrategies.fixedDelayRestart(1, 0L));
```
Are we calling `testStopWithSavepoint` with `FAIL_ON_FIRST_CHECKPOINT_ONLY`?

tillrohrmann commented on flink-tests/src/test/java/org/apache/flink/test/scheduling/DeclarativeSchedulerITCase.java:
```diff
@@ -88,6 +103,155 @@ public void testGlobalFailoverCanRecoverState() throws Exception {
+    public void testStopWithSavepointFailOnCheckpoint() throws Exception {
+        try {
+            testStopWithSavepoint(StopWithSavepointTestBehavior.FAIL_ON_CHECKPOINT);
+            fail("Expect exception");
```
Shouldn't we also assert that the job is still running?

tillrohrmann commented on flink-tests/src/test/java/org/apache/flink/test/scheduling/DeclarativeSchedulerITCase.java:
```diff
@@ -88,6 +103,155 @@ public void testGlobalFailoverCanRecoverState() throws Exception {
+            Preconditions.checkNotNull(instancesRunning);
+            instancesRunning.countDown();
+            int i = Integer.MIN_VALUE;
+            while (running) {
+                Thread.sleep(10L);
+                ctx.collect(i++);
+            }
```
I think this should be executed under the checkpoint lock.

tillrohrmann reviewed (changes requested):
Thanks for creating this PR @rmetzger. I had a couple of suggestions. The main suggestion would be to make the `StopWithSavepoint` a bit more state machine like internally instead of chaining future callbacks. Please take a look.

rmetzger commented:
@XComp I pushed a new version of this change, it is still work in progress because some tests are missing.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepoint.java:
```diff
@@ -0,0 +1,308 @@
+        final Executing.FailureResult failureResult = context.howToHandleFailure(cause);
+
+        if (failureResult.canRestart()) {
+            startCheckpointScheduler();
```
We don't have to explicitly enable Checkpoint scheduling here. This is done by the `CheckpointCoordinatorDeActivator` after the `JobStatus` changes from `RESTARTING` to `RUNNING` again.

XComp commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepoint.java:
```diff
@@ -0,0 +1,308 @@
+ * for this state is reacting to signals from the termination manager, and tracking the
+ * "operationCompletionFuture" (notify the user if we got cancelled, suspended etc. in the meantime)
+ */
+public class StopWithSavepoint extends StateWithExecutionGraph
```
I feel like we could even make `StopWithSavepoint` extend `Executing` as it's a special case of it. This way, we wouldn't need to cover certain methods like `getJobStatus` or the failure handling as it's the same as far as I can see.

rmetzger replied:
The first version of this PR did exactly this, but Till suggested to not use inheritance here, because it makes it difficult to reason about the behavior of the StopWithSavepoint state.

XComp commented on flink-tests/src/test/java/org/apache/flink/test/checkpointing/SavepointITCase.java:
```diff
@@ -684,40 +692,54 @@ private static void testStopWithFailingSourceInOnePipeline(
         }
     }
 
+    public static void waitUntilAllTasksAreRunning(URI restAddress, JobID jobId) throws Exception {
```
Good point. That improves the readability. 👍

XComp commented on flink-tests/src/test/java/org/apache/flink/test/checkpointing/SavepointITCase.java:
```diff
@@ -585,15 +585,22 @@ public void testStopWithSavepointFailingAfterSnapshotCreation() throws Exception
-                            String.format(
-                                    "Inconsistent execution state after stopping with savepoint. At least one execution is still in one of the following states: FAILED. A global fail-over is triggered to recover the job %s.",
-                                    jobId)));
+            if (isAdaptiveSchedulerEnabled(new Configuration())) {
```
I would expect the behavior being the same independent of the used scheduler.

XComp commented on flink-tests/src/test/java/org/apache/flink/test/scheduling/AdaptiveSchedulerITCase.java:
```diff
@@ -36,31 +41,48 @@
 
     private static Configuration getConfiguration() {
-        final Configuration configuration = new Configuration();
+        final Configuration conf = new Configuration();
```
I'm not sure whether the renaming is actually necessary here. I would propose reverting it for the sake of the git history.

rmetzger replied:
I renamed it because an IntelliJ inspection complained about a method field hiding a class field.

XComp commented on flink-tests/src/test/java/org/apache/flink/test/scheduling/AdaptiveSchedulerITCase.java:
```diff
@@ -88,6 +119,177 @@ public void testGlobalFailoverCanRecoverState() throws Exception {
+                    .get();
+            fail("Expect exception");
+        } catch (ExecutionException e) {
+            assertThat(e.getMessage(), containsString("Failure while stopping with savepoint"));
```
nit: `FlinkMatchers.containsMessage` might work as well

XComp commented on flink-tests/src/test/java/org/apache/flink/test/scheduling/AdaptiveSchedulerITCase.java:
```diff
@@ -88,6 +119,177 @@ public void testGlobalFailoverCanRecoverState() throws Exception {
+
+    @Test
+    public void testStopWithSavepointFailOnFirstSavepointSucceedOnSecond() throws Exception {
+        assumeTrue(ClusterOptions.isDeclarativeResourceManagementEnabled(configuration));
```
This seems to be obsolete due to `ensureDeclarativeResourceManagement()` being called before the test.a

rmetzger replied:
Too much copy-paste action. Thanks ;)

XComp commented on flink-tests/src/test/java/org/apache/flink/test/scheduling/AdaptiveSchedulerITCase.java:
```diff
@@ -72,10 +94,19 @@ private static Configuration getConfiguration() {
 
+    @Before
+    public void ensureDeclarativeResourceManagement() {
+        assumeTrue(ClusterOptions.isDeclarativeResourceManagementEnabled(configuration));
```
The `configuration` we check here is different to the one used in the `MiniCluster`. There is no real point having the member `configuration` here, I guess.

XComp replied:
```suggestion
        assumeTrue(ClusterOptions.isAdaptiveSchedulerEnabled(configuration));
```

XComp commented on flink-tests/src/test/java/org/apache/flink/test/scheduling/AdaptiveSchedulerITCase.java:
```diff
@@ -88,6 +119,177 @@ public void testGlobalFailoverCanRecoverState() throws Exception {
+        return env;
+    }
+
+    private static final class DummySource extends RichParallelSourceFunction<Integer>
```
Could we make this a generic test class that could be, for instance, also utilized in `SavepointITCase`?

rmetzger replied:
There are 200+ implementations of the SourceFunction interface in Flink, most of them in tests. It doesn't seem that there's much sharing of test sources going on.
It seems a bit arbitrary to share the sources between these two tests (feel free to point me to specific examples of the `SavepointITCase` where sharing this class makes sense)

XComp reviewed (changes requested):
Thanks @rmetzger. The changes look good. I have a few proposals which could be discussed. See my comments below.

rmetzger commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMaster.java:
```diff
@@ -430,14 +430,28 @@ protected void onStart() throws JobMasterException {
+                                        + " was not found."));
+            }
+        } catch (Throwable throwable) {
+            // if the taskExecutionState contains an error, it does not make sense to let the RPC
```
I extended this condition to catch also terminal states.

rmetzger commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StateWithExecutionGraph.java:
```diff
@@ -136,7 +136,7 @@ public ArchivedExecutionGraph getJob() {
     public void suspend(Throwable cause) {
         executionGraph.suspend(cause);
-        Preconditions.checkState(executionGraph.getState() == JobStatus.SUSPENDED);
+        // Preconditions.checkState(executionGraph.getState() == JobStatus.SUSPENDED);
```
I experienced test instabilities because of this: https://github.com/apache/flink/pull/15059

rmetzger commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveScheduler.java:
```diff
@@ -501,12 +502,11 @@ public void declineCheckpoint(DeclineCheckpoint decline) {
+                        executing -> executing.stopWithSavepoint(targetDirectory, terminate),
                         "stopWithSavepoint")
                 .orElse(
                         FutureUtils.completedExceptionally(
```
But the user requested the creation of a checkpoint, which failed because we were in the wrong state.
I believe we are using the correct exception here.

rmetzger commented:
@tillrohrmann I reworked this change. It is now based on FLINK-21030.

Tests for the StopWithSavepoint state are still missing (that's why this PR is still marked as a draft). I want to collect feedback on the approach before guarding the behavior with unit tests. (note to self: I should try to write the unit tests first to validate my assumptions)

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/stopwithsavepoint/StopWithSavepointOperationHandlerImpl.java:
```diff
@@ -49,35 +48,34 @@
-            JobID jobId, S schedulerWithCheckpointing, Logger log) {
+    public <S extends GlobalFailureHandler & StopWithSavepointOperations>
+            StopWithSavepointOperationHandlerImpl(
+                    JobID jobId, S schedulerWithCheckpointing, Logger log) {
```
Maybe update the parameter name.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/stopwithsavepoint/StopWithSavepointOperationManager.java:
```diff
@@ -18,25 +18,37 @@
+import java.util.concurrent.Executor;
 
 /**
  * {@code StopWithSavepointTerminationManager} fulfills the contract given by {@link
```
`StopWithSavepointTerminationManager` no longer exists.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/stopwithsavepoint/StopWithSavepointOperationManager.java:
```diff
@@ -45,18 +57,28 @@ public StopWithSavepointTerminationManager(
-    public CompletableFuture<String> stopWithSavepoint(
-            CompletableFuture<CompletedCheckpoint> completedSavepointFuture,
+    public CompletableFuture<String> trackStopWithSavepoint(
+            boolean terminate,
```
Does it make sense to introduce an enum to give this flag a bit more expressive power? For example only looking at the code if one passes `false` then it is not very clear to me what this means.

rmetzger replied:
Yes! I believe this is necessary. The argument order is different in various parts of the codebase, and the semantics of the boolean flag are defined different between flink-runtime and flink-core.
However, I would like to tackle this in a follow up PR, as this change alone will be difficult to review (I started doing it, but realized that this is touching a lot of classes): https://issues.apache.org/jira/browse/FLINK-21638

```
 flink-clients/src/main/java/org/apache/flink/client/deployment/application/EmbeddedJobClient.java                           |  5 +++--
 flink-core/src/main/java/org/apache/flink/core/execution/JobClient.java                                                     |  6 +++---
 flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java                                  |  6 +++---
 flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointProperties.java                                   | 17 ++++++++---------
 flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/StopWithSavepointOperations.java                            | 12 +++++++++++-
 flink-runtime/src/main/java/org/apache/flink/runtime/dispatcher/Dispatcher.java                                             |  6 ++++--
 flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMaster.java                                               |  7 +++++--
 flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMasterGateway.java                                        |  3 ++-
 flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniCluster.java                                           |  5 +++--
 flink-runtime/src/main/java/org/apache/flink/runtime/minicluster/MiniClusterJobClient.java                                  |  5 +++--
 flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/job/savepoints/SavepointHandlers.java                     |  9 +++++++--
 flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerBase.java                                           |  9 +++++----
 flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SchedulerNG.java                                             |  4 +++-
 flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/AdaptiveScheduler.java                              |  9 +++++----
 flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/Executing.java                                      |  7 ++++---
 flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepoint.java                              | 17 +++++++++--------
 flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/stopwithsavepoint/StopWithSavepointOperationManager.java     |  8 +++++---
 flink-runtime/src/main/java/org/apache/flink/runtime/webmonitor/RestfulGateway.java                                         |  5 +++--
 flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTest.java                              |  3 ++-
 flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/TestingStopWithSavepointOperations.java                     |  2 +-
 flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/ExecutingTest.java                                  |  4 +++-
 flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepointTest.java                          |  7 ++++---
 flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/stopwithsavepoint/StopWithSavepointOperationManagerTest.java |  5 +++--
 23 files changed, 99 insertions(+), 62 deletions(-)
 ```

tillrohrmann replied:
Sounds good.

tillrohrmann commented on flink-tests/src/test/java/org/apache/flink/test/scheduling/AdaptiveSchedulerITCase.java:
```diff
@@ -88,6 +121,177 @@ public void testGlobalFailoverCanRecoverState() throws Exception {
+    }
+
+    @Test
+    public void testStopWithSavepointFailOnFirstSavepointSucceedOnSecond() throws Exception {
```
Are we testing here that we clean up after a first failed savepoint operation?

tillrohrmann reviewed (commented):
Thanks for updating this PR @rmetzger. I think it looks quite good. I had a couple of comments. Please take a look.

rmetzger commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMaster.java:
```diff
@@ -430,14 +430,30 @@ protected void onStart() throws JobMasterException {
-                                    + taskExecutionState.getID()
-                                    + " was not found."));
+        try {
+            if (schedulerNG.updateTaskExecutionState(taskExecutionState)) {
```
While working on the change, I accidentally tried starting the checkpoint scheduler on a job with no checkpointing enabled in `StopWithSavepoint.updateTaskExecutionState()` (which is not the right place to do this). The condition for "no checkpointing enabled" is that the checkpointing interval is set to Long.MAX_VALUE.
If you try to start the checkpoint scheduler with that value, you end up with an integer overflow here: https://github.com/apache/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java#L1792

The exception is not caught anywhere, so the RPC system sends it back to the caller. However, in my case, the caller was notifying the JobMaster about a failure. The caller (TaskExecutor.updateTaskExecutionState()) is catching the exception coming back from the RPC call, but ignores it if the task is not anymore in the taskSlotTable. I assume that it is not in the `taskSlotTable` anymore because the task failed.

Most likely we don't have this case anywhere in our production code, but I spend quite some time figuring out this issue, because the exception from the CheckpointCoordinator doesn't get logged anywhere and is completely swallowed.
I'm not sure if it is necessary to have such defenses for these rare cases in the code, but since I experimented with this fix I left it in with a comment.

tillrohrmann replied:
In general it is ok but I would not introduce the complex logic when and when not to send the exception back. Either we fail fatally whenever an exception occurs here or we send it back and the receiver has to make sure that this is logged somehow.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepoint.java:
```diff
@@ -0,0 +1,317 @@
+                .triggerSynchronousSavepoint(terminate, targetLocation);
+    }
+
+    interface Context extends StateWithExecutionGraph.Context, GlobalFailureHandler {
```
Why do we go through the `AdaptiveScheduler` to handle global failures from the `StopWithSavepointOperationHandlerImpl`?

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepoint.java:
```diff
@@ -0,0 +1,317 @@
+        // we need to delay the execution because this may trigger a state transition of the
+        // AdaptiveScheduler.
+        context.getMainThreadExecutor()
+                .execute(
```
I think the problem you are trying to work around with this construct is that the `stopWithSavepointOperationManager.abortOperation` can trigger a `handleGlobalFailure` if it happens after the savepoint has been created. With this change we will trigger a global failure on the next state. I think this is not correct as we should only leave the `StopWithSavepoint` state if the stop with savepoint operation has been completed (one way or the other).

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/stopwithsavepoint/StopWithSavepointOperationHandlerImpl.java:
```diff
@@ -236,10 +253,17 @@ public State onExecutionsFinished() {
             return new FinalState();
         }
+
+        @Override
+        public State onAbort(Throwable cause) {
+            terminateExceptionallyWithGlobalFailover(
+                    cause, completedSavepoint.getExternalPointer());
+            return new FinalState();
+        }
```
In the context of the `StopWithSavepointOperationHandlerImpl`, this method might be correct. However, in the context of the `StopWithSavepoint` state one should use it in the `onLeave` method because it can trigger a global failover. This can only work if the state is directly responsible for handling the global failover w/o going through the `AdaptiveScheduler`. That way the `StopWithSavepoint` state could say to ignore this signal if it is about to change state.

Alternatively, a method which only fails the `result` could work. However, this would add a method which is very specific to the `StopWithSavepoint` state.

A third option would be to not directly use the `StopWithSavepointOperationHandlerImpl` directly in the `StopWithSavepoint` state if it does not fit well (at least for the case of stopping the operation and not triggering a global failure and for handling the case of a failed savepoint). Only because you have hammer does not make everything a nail.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepoint.java:
```diff
@@ -0,0 +1,317 @@
+ * delegating the tracking of the stop with savepoint operation to the {@link
+ * StopWithSavepointOperationManager}, which is shared with {@link SchedulerBase}.
+ */
+class StopWithSavepoint extends StateWithExecutionGraph implements StopWithSavepointOperations {
```
I would suggest to not let the `StopWithSavepoint` directly implement the `StopWithSavepointOperations`. That way it wouldn't be exposed to the users of the `StopWithSavepoint` class. Moreover, one could implement an inner class which controls whether to forward the calls or not depending on the internal state of `StopWithSavepoint`.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepoint.java:
```diff
@@ -0,0 +1,317 @@
+    }
+
+    @Override
+    public void startCheckpointScheduler() {
```
I think this method shouldn't be executed unconditionally. What if we are no longer in the state `StopWithSavepoint`? Then a callback from the `StopWithSavepointOperationImpl` should not have any effect.

tillrohrmann reviewed (changes requested):
Thanks for updating this PR @rmetzger. I think the PR contains a couple of problems. The main problem is that the control paths are a bit scattered between the `StopWithSavepoint` and `StopWithSavepointOperationsHandlerImpl`. I would suggest to let the `StopWithSavepoint` be in charge (ideally that this state uses the `StopWithSavepointOperationsHandlerImpl` is an implementation detail and is not visible from outside). Moreover, if the `StopWithSavepointOperationsHandlerImpl` does not work with the `StopWithSavepoint`, then it might be easier to not reuse it.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepointTest.java:
```diff
@@ -0,0 +1,458 @@
+
+        ctx.close();
+        assertThat(sws.getOperationCompletionFuture().isCompletedExceptionally(), is(true));
+        assertThat(ctx.hadStateTransition, is(false));
```
At least, we should go to `RESTARTING` if we are unsure about the execution state of the `ExecutionGraph`.

rmetzger replied:
Maybe this test should not make any assumptions about state transitions. A savepoint failure leads to a transition of the Task from RUNNING to FAILED, hence we get a notification via `updateTaskExecutionState` about this failure, which we handle accordingly.

rmetzger commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepointTest.java:
```diff
@@ -0,0 +1,458 @@
+        ctx.setExpectRestarting(assertNonNull());
+
+        sws.getSavepointFuture().complete(createCompletedSavepoint());
+        executionGraph.getExecutionFuture().complete(ExecutionState.FAILED);
```
you are absolutely right. In general, relying on the futures for detecting task failure breaks consistency because it circumvents the normal control flow of adaptive scheduler.

rmetzger commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepoint.java:
```diff
@@ -0,0 +1,317 @@
+        Preconditions.checkNotNull(
+                getExecutionGraph().getCheckpointCoordinator(),
+                "Checkpoint coordinator must be set for this operation. Is this a streaming job?");
+        return getExecutionGraph()
+                .getCheckpointCoordinator()
+                .triggerSynchronousSavepoint(terminate, targetLocation);
```
I don't understand what you are hinting at here? (note that my updated version of this PR won't expose this method as a public method of StopWithSavepoint anymore. Otherwise, the respective future completions are monitored and treated accordingly.)

tillrohrmann replied:
I can't remember either.

rmetzger commented:
@tillrohrmann Thanks a lot for your review. I haven't responded to all comments, but they were definitely helpful pushing my thinking into the right direction. I'm sorry that we have to go through so many iterations with this. 

The latest version is introducing an intermediate layer between StopWithSavepoint and StopWithSavepointOperationHandlerImpl, called StopWithSavepointOperationManagerForAdaptiveScheduler.
We could move this handler into the StopWithSavepoint class (except for the handleGlobalFailure method), as you can see from the dependencies and methods. I didn't do it yet to first collect some feedback on the overall approach.

I am aware that the test coverage is not 100% yet, but I first want approval on the approach. 
I'm 90% happy with wrapping the stop with savepoint logic of SchedulerBase. There are some cases now that we need to cover that wouldn't exist if we wouldn't reuse the logic. Maybe your happiness-ratio is lower, or you prefer 100% happiness, then I'm happy to reimplement a stop with savepoint logic for the StopWithSavepoint state.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepoint.java:
```diff
@@ -0,0 +1,264 @@
+         * @param executionGraphHandler executionGraphHandler to pass to the {@link Canceling} state
+         * @param operatorCoordinatorHandler operatorCoordinatorHandler to pass to the {@link
+         *     Canceling} state
+         */
+        void goToCanceling(
+                ExecutionGraph executionGraph,
+                ExecutionGraphHandler executionGraphHandler,
+                OperatorCoordinatorHandler operatorCoordinatorHandler);
+
+        /**
+         * Transitions into the {@link Restarting} state.
+         *
+         * @param executionGraph executionGraph to pass to the {@link Restarting} state
+         * @param executionGraphHandler executionGraphHandler to pass to the {@link Restarting}
+         *     state
+         * @param operatorCoordinatorHandler operatorCoordinatorHandler to pas to the {@link
+         *     Restarting} state
+         * @param backoffTime backoffTime to wait before transitioning to the {@link Restarting}
+         *     state
+         */
+        void goToRestarting(
+                ExecutionGraph executionGraph,
+                ExecutionGraphHandler executionGraphHandler,
+                OperatorCoordinatorHandler operatorCoordinatorHandler,
+                Duration backoffTime);
+
+        /**
+         * Transitions into the {@link Failing} state.
+         *
+         * @param executionGraph executionGraph to pass to the {@link Failing} state
+         * @param executionGraphHandler executionGraphHandler to pass to the {@link Failing} state
+         * @param operatorCoordinatorHandler operatorCoordinatorHandler to pass to the {@link
+         *     Failing} state
+         * @param failureCause failureCause describing why the job execution failed
+         */
+        void goToFailing(
+                ExecutionGraph executionGraph,
+                ExecutionGraphHandler executionGraphHandler,
+                OperatorCoordinatorHandler operatorCoordinatorHandler,
+                Throwable failureCause);
```
How are we going back to `Executing` if the savepoint future failed? Or can this not happen?

rmetzger replied:
My assumption while writing the code was that a savepoint failure leads to a transition of the Task from RUNNING to FAILED, hence we get a notification via updateTaskExecutionState about this failure, where we then go into the Failing state.

However, I see that there might be a scenario where the Checkpoint coordinator notifies us (via the savepoint future) that the savepoint failed, before we receive the notification from the TaskExecutor that the Task failed. In this case, we would indeed be stuck in the StopWithSavepoint state, somehow depending on the Task failure to get us out there.
Also, if the implementation of the Task changes, we will have to touch the StopWithSavepoint state.

I'm probably a bit too test test driven in my development ;)

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepointOperationManagerForAdaptiveScheduler.java:
```diff
@@ -0,0 +1,117 @@
+
+        FutureUtils.assertNoException(
+                savepointFuture
+                        // the completedSavepointFuture could also be completed by
+                        // CheckpointCanceller which doesn't run in the mainThreadExecutor
```
This is a very good finding.

rmetzger replied:
It's copy-pasted from Matthias implementation ;)

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepointOperationManagerForAdaptiveScheduler.java:
```diff
@@ -0,0 +1,117 @@
+                        executionGraph.getJobID(), this, stopWithSavepointOperations, logger);
+
+        FutureUtils.assertNoException(
+                savepointFuture
+                        // the completedSavepointFuture could also be completed by
+                        // CheckpointCanceller which doesn't run in the mainThreadExecutor
+                        .handleAsync(
+                        (completedSavepoint, throwable) -> {
+                            handler.handleSavepointCreation(completedSavepoint, throwable);
+                            return null;
+                        },
+                        mainThreadExecutor));
```
We could also remove this special handling here by saying that `StopWithSavepointOperations.triggerSynchronousSavepoint` must return a future which is completed in the main thread. That way we move the problem out of this class. This reduces a bit of complexity here.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepointOperationManagerForAdaptiveScheduler.java:
```diff
@@ -0,0 +1,117 @@
+        // state, the handler will complete the stop with savepoint future
+        // if we are in any error state, the StopWithSavepoint state will decide what to do.
+        operationFinished = true;
+        Collection<ExecutionState> executionsState =
+                executionGraph.getRegisteredExecutions().values().stream()
+                        .map(Execution::getState)
+                        .collect(Collectors.toList());
+        handler.handleExecutionsTermination(executionsState);
```
Intuitively, I would have expected an `handler.abort` call here instead.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/stopwithsavepoint/StopWithSavepointOperationHandlerImpl.java:
```diff
@@ -236,10 +260,17 @@ public State onExecutionsFinished() {
+
+        @Override
+        public State onAbort(Throwable cause) {
+            terminateExceptionallyWithGlobalFailover(
+                    cause, completedSavepoint.getExternalPointer());
```
Why do we fail globally `onAbort`? Wouldn't it make more sense to say that `onAbort` stops the stop with savepoint operation but the caller has to make sure that the system continues correctly?

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/stopwithsavepoint/StopWithSavepointOperationHandlerImpl.java:
```diff
@@ -217,6 +230,17 @@ public State onSavepointCreationFailure(Throwable throwable) {
+
+        @Override
+        public State onAbort(Throwable cause) {
+            return onSavepointCreationFailure(cause);
```
Why does `onAbort` trigger a savepoint creation failure? Couldn't it transition to `Finished` and fail the operation future?

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepointOperationManagerForAdaptiveScheduler.java:
```diff
@@ -0,0 +1,117 @@
+
+    @Override
+    public void handleGlobalFailure(Throwable cause) {
+        // ignore failures reported from the StopWithSavepointOperationHandlerImpl.
+        Preconditions.checkState(
+                operationFinished,
+                "Unexpected failure. Assuming failures only after operation has finished.");
```
It is a bit confusing that we only expect global failures to happen after the operation is finished. I took me a while to understand the reason tbh.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepointOperationManagerForAdaptiveScheduler.java:
```diff
@@ -0,0 +1,117 @@
+                        // CheckpointCanceller which doesn't run in the mainThreadExecutor
+                        .handleAsync(
+                        (completedSavepoint, throwable) -> {
+                            handler.handleSavepointCreation(completedSavepoint, throwable);
```
I think we will stay forever in the state `StopWithSavepoint` if the savepoint future fails.

tillrohrmann reviewed (changes requested):
Thanks for updating the PR @rmetzger. I am not sure whether the `StopWithSavepointOperationManagerForAdaptiveScheduler` is really what we wanna have. There is at least concrete problem which is that we get stuck in the `StopWithSavepoint` state if the savepoint future fails.

Let me maybe try to write a rough draft of how one could do it.

tillrohrmann commented:
Here is [a branch with a simple `StopWithSavepoint` implementation](https://github.com/tillrohrmann/flink/tree/FLINK-21333). Let me know what you think. This can definitely be more beautified and I might have overlooked some cases.

rmetzger commented:
Thanks a lot for your draft! It's impressive how concise you've expressed the problem! I didn't find a lot of opportunities to beautify your code, but I extended the tests a bit more and massively reduced the number of changes on this PR.

Looking forward to your feedback!

rmetzger commented on flink-tests/src/test/java/org/apache/flink/test/scheduling/AdaptiveSchedulerITCase.java:
```diff
@@ -88,6 +123,189 @@ public void testGlobalFailoverCanRecoverState() throws Exception {
+    }
+
+    @Test
+    public void testStopWithSavepointFailOnCheckpoint() throws Exception {
```
I will add such a test

rmetzger replied:
I didn't find a nice way (without a lot of effort of sneaking in a mocked Checkpoint coordinator) to add this IT case.
Are you okay with testing this case only in a unit test?

tillrohrmann replied:
Yes, unit test is good enough.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepoint.java:
```diff
@@ -0,0 +1,328 @@
+                            context.runIfState(
+                                    this,
+                                    () -> handleSavepointCompletion(savepointLocation, throwable),
+                                    Duration.ZERO);
```
Maybe add a comment stating that we need this in order to not execute the callback within the constructor.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepoint.java:
```diff
@@ -0,0 +1,328 @@
+                completeOperationAndGoToFinished(savepoint);
+            }
+        } else {
+            handleAnyFailure(new FlinkException("Job did not finish properly."));
```
```suggestion
            handleAnyFailure(new FlinkException("Job did not reach the FINISHED state while performing stop-with-savepoint."));
```

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepoint.java:
```diff
@@ -0,0 +1,328 @@
+ * considered successfully if the "savepointFuture" completed successfully, and the job reached the
+ * terminal state FINISHED.
+ */
+class StopWithSavepoint extends StateWithExecutionGraph {
```
Where are we stopping the `CheckpointCoordinator`? I guess we are lacking test coverage for this.

rmetzger replied:
Very good catch! Thanks a lot!

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/ExecutingTest.java:
```diff
@@ -246,7 +251,30 @@ public void testFalseReportsViaUpdateTaskExecutionStateAreIgnored() throws Excep
+                            .build(ctx);
+
+            ctx.setExpectStopWithSavepoint(assertNonNull());
+            exec.stopWithSavepoint("file:///tmp/target", true);
```
We should check that the checkpoint scheduling has been stopped.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepointTest.java:
```diff
@@ -0,0 +1,491 @@
+            ctx.setExpectFinished(assertNonNull());
+
+            savepointFuture.complete(SAVEPOINT_PATH);
+            mockExecutionGraph.completeTerminationFuture(JobStatus.FINISHED);
+            ctx.triggerScheduledExecutors();
```
Are testing the different orderings of signals? 

1) First savepoint future completion and then final state
2) First final state and then savepoint future completion

tillrohrmann replied:
Are you sure that we don't have the same ordering as in `testJobFinishedBeforeSavepointFuture`?

rmetzger replied:
in `testJobFinishedBeforeSavepointFuture`, I test the 2nd case., in `testFinishedOnSuccessfulStopWithSavepoint` I test the 1st case.

tillrohrmann replied:
I think this is not correct. Take a look at when which callback is called in the `StopWithSavepoint` implementation.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepointTest.java:
```diff
@@ -0,0 +1,491 @@
+
+    @Test
+    public void testJobFailed() throws Exception {
+        try (MockStopWithSavepointContext ctx = new MockStopWithSavepointContext()) {
+            StateTrackingMockExecutionGraph mockExecutionGraph =
+                    new StateTrackingMockExecutionGraph();
+            StopWithSavepoint sws = createStopWithSavepoint(ctx, mockExecutionGraph);
+            ctx.setStopWithSavepoint(sws);
+            ctx.setHowToHandleFailure(Executing.FailureResult::canNotRestart);
+
+            ctx.setExpectFailing(
+                    failingArguments -> {
+                        assertThat(
+                                failingArguments.getExecutionGraph().getState(),
+                                is(JobStatus.FAILED));
+                        assertThat(
+                                failingArguments.getFailureCause(),
+                                containsCause(FlinkException.class));
+                    });
+
+            // fail job:
+            mockExecutionGraph.completeTerminationFuture(JobStatus.FAILED);
+            ctx.triggerScheduledExecutors();
+
+            assertThat(sws.getOperationFuture().isCompletedExceptionally(), is(true));
+        }
```
Does this test completes the savepoint operation? This would be good to know for the test whether the state waits for it or not.

rmetzger replied:
I'm not sure if I fully understood this comment. I've added `testJobFailedAndSavepointOperationFails`, hoping to address your concern.

tillrohrmann replied:
The question was whether the savepoint future is completed or not. Looking at this test, this is not clear to me. I guess this is somewhere hidden in the default values when creating `createStopWithSavepoint`.

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepointTest.java:
```diff
@@ -0,0 +1,491 @@
+            ctx.setStopWithSavepoint(sws);
+            ctx.setExpectCancelling(assertNonNull());
+
+            sws.cancel();
```
We should check that the operation has been failed.

tillrohrmann replied:
Fair enough. We have tested this with `testExceptionalFutureCompletionOnLeaveWhileWaitingOnSavepointCompletion`

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepointTest.java:
```diff
@@ -0,0 +1,491 @@
+                        assertThat(archivedExecutionGraph.getState(), is(JobStatus.SUSPENDED));
+                    });
+
+            sws.suspend(new RuntimeException());
```
Same here.

tillrohrmann replied:
Covered by `testExceptionalFutureCompletionOnLeaveWhileWaitingOnSavepointCompletion`

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepointTest.java:
```diff
@@ -0,0 +1,491 @@
+
+            ctx.setExpectRestarting(assertNonNull());
+
+            sws.handleGlobalFailure(new RuntimeException());
```
Same here.

tillrohrmann replied:
Covered by `testExceptionalFutureCompletionOnLeaveWhileWaitingOnSavepointCompletion`

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepointTest.java:
```diff
@@ -0,0 +1,491 @@
+                                containsCause(RuntimeException.class));
+                    });
+
+            sws.handleGlobalFailure(new RuntimeException());
```
Same here.

tillrohrmann replied:
Covered by `testExceptionalFutureCompletionOnLeaveWhileWaitingOnSavepointCompletion`

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepointTest.java:
```diff
@@ -0,0 +1,491 @@
+    }
+
+    @Test
+    public void testExceptionalFutureCompletionOnLeaveWhileWaitingOnSavepointCompletion()
```
```suggestion
    public void testExceptionalOperationFutureCompletionOnLeaveWhileWaitingOnSavepointCompletion()
```

tillrohrmann reviewed (changes requested):
Thanks for updating this PR @rmetzger. The changes look already really good. I had a couple of comments. The main one is probably that we don't seem to stop the scheduling of checkpoints when going into the `StopWithSavepoint` state.

rmetzger commented:
Thanks a lot for your feedback! I hope I addressed all your comments.
I also cleaned up the history (only the last round of review is in a separate commit)

tillrohrmann commented on flink-runtime/src/test/java/org/apache/flink/runtime/scheduler/adaptive/StopWithSavepointTest.java:
```diff
@@ -0,0 +1,535 @@
+    private static final String SAVEPOINT_PATH = "test://savepoint/path";
+
+    @Test
+    public void testFinishedOnSuccessfulStopWithSavepoint() throws Exception {
+        try (MockStopWithSavepointContext ctx = new MockStopWithSavepointContext()) {
+            StateTrackingMockExecutionGraph mockExecutionGraph =
+                    new StateTrackingMockExecutionGraph();
+            CompletableFuture<String> savepointFuture = new CompletableFuture<>();
+
+            StopWithSavepoint sws =
+                    createStopWithSavepoint(ctx, mockExecutionGraph, savepointFuture);
+            ctx.setStopWithSavepoint(sws);
+            ctx.setExpectFinished(assertNonNull());
+
+            savepointFuture.complete(SAVEPOINT_PATH);
+            mockExecutionGraph.completeTerminationFuture(JobStatus.FINISHED);
+            ctx.triggerExecutors();
+
+            assertThat(sws.getOperationFuture().get(), is(SAVEPOINT_PATH));
+        }
+    }
```
I think the execution order is that `StopWithSavepoint` first runs `onGloballyTerminalState` and then `handleSavepointCompletion`. This is the similar order as in `testJobFinishedBeforeSavepointFuture`.

tillrohrmann reviewed (approved):
Thanks for updating this PR @rmetzger. I think the changes look good. One open comment is the test case and the order in which the savepoint future and the `onGloballyTerminalState` is executed. I think `StopWithSavepointTest.testFinishedOnSuccessfulStopWithSavepoint` is not correct. +1 for merging this PR after resolving my comments.

rmetzger commented:
Thanks a lot for your review & approval. I'll merge this PR with your last comments addressed once CI has passed.
